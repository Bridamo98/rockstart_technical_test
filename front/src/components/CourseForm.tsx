import { zodResolver } from "@hookform/resolvers/zod";
import { useEffect, useState } from "react";
import { useForm } from "react-hook-form";
import * as z from "zod";

import { InstructorsAPI } from "../api/instructors";
import type { Instructor } from "../types";

const createSchema = z.object({
  title: z
    .string()
    .min(1, "Title is required")
    .max(255, "Title must be â‰¤ 255 characters"),
  course_desc: z.string().min(1, "Description is required"),
  instructor_id: z.number().int().positive("Select instructor"),
});
const updateSchema = createSchema.partial();

type CreateFormData = z.infer<typeof createSchema>;
type UpdateFormData = z.infer<typeof updateSchema>;
type FormData = CreateFormData | UpdateFormData;

interface Props {
  defaultValues?: UpdateFormData;
  onSubmit: (values: FormData) => void | Promise<void>;
  submitLabel?: string;
}

export default function CourseForm({
  defaultValues,
  onSubmit,
  submitLabel = "Save",
}: Props) {
  const isEdit = !!defaultValues;
  const schema = isEdit ? updateSchema : createSchema;

  const {
    register,
    handleSubmit,
    setValue,
    formState: { errors },
  } = useForm<FormData>({
    defaultValues,
    resolver: zodResolver(schema),
  });

  const [instructors, setInstructors] = useState<Instructor[]>([]);

  useEffect(() => {
    InstructorsAPI.list().then((r) => {
      setInstructors(r.data);
      if (!defaultValues?.instructor_id && r.data.length > 0) {
        setValue("instructor_id", r.data[0].id);
      }
    });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <form
      onSubmit={handleSubmit(onSubmit)}
      className="space-y-4 bg-white p-4 rounded shadow"
    >
      <div>
        <input
          {...register("title")}
          maxLength={255}
          placeholder="Title"
          className="input w-full"
        />
        {errors.title && (
          <p className="text-sm text-red-600">{errors.title.message}</p>
        )}
      </div>
      <div>
        <textarea
          {...register("course_desc")}
          placeholder="Description"
          className="textarea w-full"
        />
        {errors.course_desc && (
          <p className="text-sm text-red-600">{errors.course_desc.message}</p>
        )}
      </div>
      <div>
        <select {...register("instructor_id")} className="select w-full">
          {instructors.map((i) => (
            <option key={i.id} value={i.id}>
              {i.name}
            </option>
          ))}
        </select>
        {errors.instructor_id && (
          <p className="text-sm text-red-600">{errors.instructor_id.message}</p>
        )}
      </div>

      <button className="btn btn-primary">{submitLabel}</button>
    </form>
  );
}
