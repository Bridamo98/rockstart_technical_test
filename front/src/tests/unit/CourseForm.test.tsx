import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { expect, test, vi } from "vitest";
import CourseForm from "../../components/CourseForm";

test("UI inicial coincide con snapshot", () => {
  const { asFragment } = render(<CourseForm onSubmit={vi.fn()} />);
  expect(asFragment()).toMatchSnapshot();
});

test("muestra errores si se envía vacío", async () => {
  const onSubmit = vi.fn();
  const { asFragment } = render(<CourseForm onSubmit={onSubmit} />);
  const user = userEvent.setup();

  await user.click(screen.getByRole("button", { name: /save/i }));

  expect(await screen.findByText(/title is required/i)).toBeInTheDocument();
  expect(asFragment()).toMatchSnapshot("CourseForm-con-errores");
  expect(onSubmit).not.toHaveBeenCalled();
});
