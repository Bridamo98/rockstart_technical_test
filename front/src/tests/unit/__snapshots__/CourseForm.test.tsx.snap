// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UI inicial coincide con snapshot 1`] = `
<DocumentFragment>
  <form
    class="space-y-4 bg-white p-4 rounded shadow"
  >
    <div>
      <input
        class="input w-full"
        maxlength="255"
        name="title"
        placeholder="Title"
      />
    </div>
    <div>
      <textarea
        class="textarea w-full"
        name="course_desc"
        placeholder="Description"
      />
    </div>
    <div>
      <select
        class="select w-full"
        name="instructor_id"
      />
    </div>
    <button
      class="btn btn-primary"
    >
      Save
    </button>
  </form>
</DocumentFragment>
`;

exports[`muestra errores si se envía vacío > CourseForm-con-errores 1`] = `
<DocumentFragment>
  <form
    class="space-y-4 bg-white p-4 rounded shadow"
  >
    <div>
      <input
        class="input w-full"
        maxlength="255"
        name="title"
        placeholder="Title"
      />
      <p
        class="text-sm text-red-600"
      >
        Title is required
      </p>
    </div>
    <div>
      <textarea
        class="textarea w-full"
        name="course_desc"
        placeholder="Description"
      />
      <p
        class="text-sm text-red-600"
      >
        Description is required
      </p>
    </div>
    <div>
      <select
        class="select w-full"
        name="instructor_id"
      >
        <option
          value="1"
        >
          Inst
        </option>
      </select>
    </div>
    <button
      class="btn btn-primary"
    >
      Save
    </button>
  </form>
</DocumentFragment>
`;
